packages/sanity/src/core/form/inputs/files/ImageInput/usePreviewImageSource.ts

```diff
@@ -0,0 +1,62 @@
+import {isImageSource} from '@sanity/asset-utils'
+import {type CSSProperties, useMemo} from 'react'
+import {useDevicePixelRatio} from 'use-device-pixel-ratio'
+
+import {type ImageUrlBuilder} from '../types'
+import {type BaseImageInputValue} from './types'
+
+type Dimensions = [width: number, height: number]
+
+/*
+  Used for setting the initial image height - specifically for images
+  that are small and so can take less space in the document
+*/
+const getImageSize = (src: string): Dimensions => {
+  const imageUrlParams = new URLSearchParams(src.split('?')[1])
+  const rect = imageUrlParams.get('rect')
+
+  if (rect) {
+    return [rect.split(',')[2], rect.split(',')[3]].map(Number) as Dimensions
+  }
+
+  return src.split('-')[1].split('.')[0].split('x').map(Number) as Dimensions
+}
+
+export function usePreviewImageSource<Value extends BaseImageInputValue | undefined>({
```

Comment:
is there any value in this being generic? i.e. are there known cases where we'd like to preserve type information between argument type and return type when calling this function? (asking out of curiosity)

Reply:
Only as a nicety for typing the return value. If the input `value` is narrowed to exclude `undefined`, then the output `url` will always be `string`—so the consumer doesn't have to check it again.

For example, where we use it here:

<img width="594" alt="Screenshot 2024-10-01 at 09 31 03" src="https://github.com/user-attachments/assets/2a69e122-381e-4224-8030-4143a02ae6b1">

However, if `value` might be `undefined`:

<img width="757" alt="Screenshot 2024-10-01 at 09 30 41" src="https://github.com/user-attachments/assets/798c0177-412a-4380-9689-66504b82cd55">

So, it's not strictly necessary. Happy to remove it if we think it's too complicated for the value it provides.

Reply:
Cool, makes sense